ebpf/aya-ebpf/src/lib.rs

```diff
@@ -133,6 +133,28 @@ pub fn check_bounds_signed(value: i64, lower: i64, upper: i64) -> bool {
     }
 }
 
+#[macro_export]
+macro_rules! main_stub {
+    () => {
+        #[cfg(not(target_arch = "bpf"))]
+        fn main() {
+            panic!(r#"eBPF kernels are not designed to be executed in user-space. This main function is only a placeholder to allow the code to compile on the host system (i.e. on any system that is not `target_arch = "bpf"`). This works in tandem with the `no_main` attribute which is only applied when compiling for `target_arch = "bpf"`."#)
+        }
+    };
+}
+
+/// Defines our panic handler when compiling for eBPF.
+///
+/// eBPF programs are not allowed to panic, meaning this handler won't actually ever be called.
```

Comment:
This is incorrect. It's not that BPF programs aren't allowed to panic, it is that there's no concept of panic in BPF.

We use a panic handler implemented as an infinite loop because the verifier disallows non-terminating loops. That guarantees that if any panics survive in the generated code, the verifier will reject the resulting program (because it would be an infinite loop).

Reply:
I've mentioned this in passing in other convos, but an unwinding panic handler could be implemented with special glue in the userspace loader code, where the loader would hot patch at program load time the panic code to unwind the stack and return some panic-default value (and perhaps print an error with aya-log).

Low on my priority list, but at least in theory we have the technology (tm)